jax/interpreters/masking.py

```diff
@@ -403,6 +403,7 @@ def process_primitive(self, primitive, tracers, params):
     out_aval = primitive.abstract_eval(*(t.aval for t in tracers), **params)
     vals, polymorphic_shapes = unzip2((t.val, t.polymorphic_shape) for t in tracers)
     logical_shapes = map(shape_as_value, polymorphic_shapes)
+    if primitive.name == 'reshape': params['polymorphic_shapes'] = polymorphic_shapes
```

Comment:
We should probably avoid this kind of pattern. I think the general form of a masking rule should be to pass in polymorphic shapes, and if the common case doesn't need it we can just have convenience wrappers to handle those.

I can make that refactor, but can you leave a "TODO(mattjj): generalize mask rule signature" above this line?